[영화 극장 선택] 악어 1, 2단계 미션 제출합니다.

Hogu59 opened the pull request:
제리 안녕하세요!
처음 뵙습니다. 악어입니다! 🐊 

## 질문
1. Fragment UI 테스트)
- 버튼을 눌렀을 때 bottomSheetDialog Fragment가 뜨는데, 이러한 부분도 해당 Fragment 에서 UI 테스트를 진행해야 하는 것인지 궁금합니다! fragment전환 역시도 테스트를 보통 하는지 궁금합니다.

2. 액티비티 전환 테스트)
- Activity 간의 전환 테스트는 ui-test의 대상이 아니라는 리뷰가 지난번 미션에서 유행했었는데, A 액티비티에서 B 액티비티로 이동할지, 혹은 어떤 액티비티로 이동할지에 대해서 테스트를 하지 않는다면 전환 된 내용에 대한 보장된 테스트가 없으므로, 테스트가 충분하지 않다고 생각합니다. 현업에서도 activity 간의 전환되는 부분을 테스트 하지 않는지, 아니면 다른 방식으로 테스트를 수행하는 지 알고 싶습니다.

[페어와 작업한 내용입니다]
https://github.com/Hogu59/android-movie-theater/compare/64064917630d96bd111cadaeb7b80da0b9041a05..3967d9ef5539da1453ff309155bfae59f7c5eac6

[Screen_recording_20240502_162728_out.webm](https://github.com/woowacourse/android-movie-theater/assets/62333909/2cacd151-9c7b-4920-bf00-043d07efe826)

리뷰 잘 부탁드립니다! 🙇

Hogu59 committed:
initialize project

Hogu59 committed:
docs: 기능 목록 작성

Hogu59 committed:
feat(Screen): 영화 상영 정보를 담는 클래스 구현

Hogu59 committed:
feat(Theater): 영화관 정보를 담는 클래스 구현

Hogu59 committed:
feat(TheaterRepository): 극장 더미 데이터 생성

Hogu59 committed:
feat(BottomSheetFragment): 극장 선택 fragment 구현

Hogu59 committed:
refactor(movieList): viewholder 패키지 분리

Hogu59 committed:
feat(movieList): 더미 데이터와 상영 정보 바인딩

Hogu59 committed:
feat(movieList): 화면 이동 구현
- from: MovieListActivity
- to: TicketingActivity

Hogu59 committed:
refactor: databinding 적용 (on process)

- 액티비티 값 넘겨주는 내용 진행중
- !!해당 커밋으로 이동 금지!!

Hogu59 committed:
feat(movieList): 액티비티 이동 간 데이터 전달 구현
- theaterId 전달 구현 (마지막 액티비티까지)

Hogu59 committed:
style: 패키지 분리

Hogu59 committed:
feat: 영화 상영 시간의 개수 출력 구현

Hogu59 committed:
feat(TicketingResult): 티켓 정보 출력 구현
- 티켓 정보 데이터 바인딩 구현
- Ticket Class
  - priceToString 함수 구현
  - seatToString 함수 구현
- TheaterListRepository
  - findTheaterName 함수 구현

Hogu59 committed:
feat(result.xml): 극장 이름 길이에 따른 말줄임표 구현

Hogu59 committed:
feat: 극장/영화에 따른 시간 스피너 변동 구현

Hogu59 committed:
feat: 네비게이션바 구현

Hogu59 committed:
refactor: 데이터 바인딩 적용

Hogu59 committed:
refactor(SeatSelection): 데이터 바인딩 적용

Hogu59 committed:
refactor(*): Apply Databinding

Hogu59 committed:
test(homeUI): 초기 화면 테스트 작성

Hogu59 committed:
test(ticketingUI): Ticketing UI 테스트 작성
- Ticketing: 영화 상세 정보 표시 페이지

Hogu59 committed:
test(ticketingResultUI): Ticketing Result UI 테스트 작성
- DummyTheaterList 에 긴 극장 이름 추가

Hogu59 committed:
test(MovieListFragment): 테스트 코드 작성

vagabond95 commented:
작업한 커밋 범위로 공유 주셔서 넘 좋네요 👍 

> Fragment UI 테스트, 액티비티 전환 테스트

둘다 같은 맥락으로 얘기드릴 수 있을 것 같네요.

여러 리뷰이분들에게 우테코 교육방향에 반하는 내용을 전파하고 다니는 것 같아 뜨끔하지만 ^^;
저는 기본적으로 전반적인 UI 테스트에 대해 회의적인 입장입니다. (unit test 는 👍 )

작업 대비 효용성이 너무 떨어진다는 이유 때문입니다.
`화면 전환이 되는 것을 검증하자!` 라고 목표를 세운다면, 모든 화면 시나리오 flow 에 대한 테스트 코드를 작성하고 처리해야하는데 프로젝트가 커지고 제한된 일정내에서 이것을 고려하며 개발하기란 너무 어려운 것이 현실적인 상황입니다.

그럼에도 불구하고 나는 화면 flow 를 검증하고 싶다. 라고 한다면,
MVP 패턴으로 잘 구현이 되었다는 가정하에, 화면간 전환을 명령하는 presenter 함수의 호출 여부를 검증해볼 수는 있을 것 같습니다.

테스트 유무에 대해 질문 주셨는데, "나는 테스트 안한다!" 로 답변 드려서 당황스러우실수도 있지만, 이또한 하나의 의견으로 받아들여주시면 좋겠습니다 :)

vagabond95 commented on app/src/main/res/layout-land/activity_ticketing.xml:
```diff
@@ -0,0 +1,235 @@
+<?xml version="1.0" encoding="utf-8"?>
```
landscape 용 xml 을 별도로 만들어주셨군요 👍

vagabond95 commented on app/src/main/java/woowacourse/movie/utils/FormatSeat.kt:
```diff
@@ -0,0 +1,5 @@
+
+import woowacourse.movie.model.Seat
+
+fun formatSeat(seat: Seat): String = "${('A'.code + seat.row).toChar()}${seat.col + 1}"
```
데이터를 mapping 하는 함수는 extension function 을 활용하면 가독성도 살리고 좀 더 자연스러운 사용이 가능합니다!

```
"Text".toSeat()
```

Hogu59 replied:
그렇네요! 확장함수로 domain 안에서 구현하여 사용하면, 함수의 위치가 더 자연스러울 것 같네요!
특히나 utils 패키지를 만들 필요가 없앨 수 있을 것 같습니다. 감사합니다 😄

vagabond95 commented on app/src/main/res/layout/fragment_theater_bottom_sheet_dialog.xml:
```diff
@@ -0,0 +1,19 @@
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
+            tools:listitem="@layout/item_theater" />
```
tools:listitem 활용 👍

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/home/HomeActivity.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private fun setupInitialFragment() {
+        supportFragmentManager.beginTransaction()
+            .add(R.id.fragment_container, MovieListFragment())
+            .commit()
```
fragment transaction 은 항상 마지막에 commit() 함수를 호출해줘야 내용이 반영되는데요. 이를 누락하는 실수를 줄이고자 + 편의성을 위해 commit { } 함수를 제공하고 있으니 활용해보셔도 좋겠습니다 :)

Hogu59 replied:
오 좋은 방법이네요! 몰랐었는데 `commit { }`을 활용하면 확실히 편하면서도 commit() 누락을 피할 수 있겠네요!
알려주셔서 감사합니다 😄

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/home/HomeActivity.kt:
```diff
@@ -0,0 +1,64 @@
+
+    private fun replaceFragment(fragment: Fragment) {
+        supportFragmentManager.commit {
+            setReorderingAllowed(true)
+            replace(R.id.fragment_container, fragment)
+            addToBackStack(null)
```
setReorderingAllowed, addToBackStack 옵션을 넣어주신 이유를 알 수 있을까요?

Hogu59 replied:
앗.. 사실 안드로이드에 대한 지식이 많지 않아 예시코드를 그대로 사용했었습니다!
잘 모르고 보일러플레이트처럼 있는 건줄 알고 그냥 사용했었는데, 덕분에 찾아보게 되었습니다 감사합니다 🙇 

1. serReorderingAllowed
 - [developer.android 문서](https://developer.android.com/reference/androidx/fragment/app/FragmentTransaction?_gl=1*kavomp*_up*MQ..*_ga*NzkwNjkwNzAuMTcxNDg4MzY3Mg..*_ga_6HH9YJMN9M*MTcxNDg4MzY3MS4xLjAuMTcxNDg4MzY3MS4wLjAuMA..#setReorderingAllowed(boolean)) 를 참고해 공부해보았는데, fragment의 불필요한 호출을 최적화해주는 역할을 하는 코드네요. 현재 미션에서는 두 개 이상의 fragment가 중복 호출되어 비효율을 초래할 상황은 거의 없을 것 같지만, 프로젝트가 충분히 크다면 혹시 모를 문제를 대비해 습관적으로 사용하는 것이 조금 더 좋을 것 같다는 생각이 드네요. (비록 보일러 플레이트 같아질 여지가 많아보이기도 합니다만..)
 - A와 B 두 개의 fragment를 동시에 호출하는 코드에서, B가 실행되기 전에 A의 생성이나 동작이 꼭 필요한 상황을 제외하고서는 최적화해주는 해당 코드를 사용하는 게 바람직 할 것 같다는 생각이 드네요... 🤔 
 - 혹시 제리의 현업에서는 `setReorderingAllowed`를 사용하시는지, 혹시 제가 생각한 방법이 아닌 다른 의도로 사용하신다면 혹시 다른 의도는 무엇인지 알 수 있을까요?

2. addToBackStack
 - [developer.android 문서](https://developer.android.com/guide/fragments/fragmentmanager) 를 참고할 때, 뒤로가기 버튼을 사용하여 이전 fragment를 사용할 목적이 아니라면 해당 메서드는 사용하지 않는 것이 좋을 것 같네요!

모르는 코드들을 그저 사용했던 것 같네요..! 알려주셔서 감사합니다! 😄

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/home/HomeActivity.kt:
```diff
@@ -0,0 +1,64 @@
+    private fun replaceFragment(fragment: Fragment) {
+        supportFragmentManager.commit {
+            setReorderingAllowed(true)
+            replace(R.id.fragment_container, fragment)
```
replace 시, tag 를 넣어주는 습관을 들이시면 좋습니다.
현 단계에서는 크게 활용할 일이 없으나, 추후에 특정 fragment 를 조회하고 이를 이용한 처리를 할 때 필수적인 값으로 활용이 됩니다!

Hogu59 replied:
오.. `add` 나 `replace` 메서드를 사용할 때는 tag를 활용할 수 있군요!
확실히 tag 를 항상 넣어주면(네이밍이 항상 적절하다는 전제하에), `findFragmentByTag`등을 활용하는데 용이할 것 같네요! 알려주셔서 감사합니다! 😄

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/homefragments/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,46 @@
+
+    override fun ticketingButtonClick(movieId: Long) {
+        val bottomSheet = TheaterBottomDialogFragment(DummyTheaterList, movieId)
+        bottomSheet.show(parentFragmentManager, bottomSheet.tag)
```
parentFragmentManager 와 childFragmentManager 를 전달했을 때의 차이는 무엇이 있을까요?

Hogu59 replied:
bottomSheetDialogFragment의 생명주기가 각각 다른 생명주기를 따르게 될 것 같습니다.
1) `parentFragmentManager` 를 넘겨줄 경우, MovieListFragment의 상위인 HomeActivity의 생명주기를 따르게 되며,
2) `childFragmentManager` 를 넘겨줄 경우, MovieListFragment의 생명주기를 따르게 될 것 같습니다.
이런 맥락에서는 TheaterBottomDialogFragment 가 activity 의 생명주기보다 MovieListFragment의 생명주기를 따르는 것이 자연스러우므로, childFragmentManager를 넘겨주는 것이 더 좋을 것 같습니다.

다만, Log를 찍어보면서 확인을 해볼때는,
bottomSheetDialogFragment 에서 다음 acitivity인 TicketingActivity로 넘어가더라도
parentFragmentManager를 넘겨주든 childFragmentManager를 넘겨주든 생명주기가 똑같이 수행되어 무엇이 실제로 다른지 차이를 확인하기가 어려웠습니다.

그렇다면,
1) 제가 fragmentManager에 대해 이해한 내용이 잘못되었거나, 
2) 질문의 요지를 잘못 이해했거나,
3) 현재의 프로젝트에서 넘겨주는 인자가 다를 때 차이점을 확인하기 위한 방법은 다른 방법으로 수행되어야 하는 경우가 있을 것 같은데,

어떻게 다른지 확인할 방법을 좀 알 수 있을까요...? 로그를 다 찍어봐도 차이점이 없다보니 어떻게 차이가 있는지를 실제로 체감하기가 조금 어려운 것 같습니다... 😢

vagabond95 replied:
해당 코멘트를 남긴 의도는 대부분의 리뷰이가 parentFragmentManager 를 무의식중으로 사용하시는 것 같아 남기게 되었습니다.
childFragmentManager 라는 또다른 FragmentManager 가 있는데, 각각의 구분지어놨다는 것은 서로 다른 역할을 수행할 수 있다는 것이므로 공부해보면 좋을 주제같아 보여서요 :)

힌트를 조금 드리자면 FragmentManager 는 Fragment 의 생명주기와는 무관하며, backstack 과 깊은 연관이 있습니다.
아래 두 문서를 참고해보시면 좋을 것 같습니다 :)

- [링크1](https://developer.android.com/guide/fragments/fragmentmanager?hl=ko#considerations)
- [링크2](https://medium.com/shdev/fragment-back-stack%EC%9D%98-%EC%9D%B4%ED%95%B4-ed30638a57f1)

vagabond95 commented on app/src/main/java/woowacourse/movie/model/Theater.kt:
```diff
@@ -0,0 +1,14 @@
+
+import java.time.LocalTime
+
+class Theater(val id: Long, val name: String, val screens: List<Screen>) {
```
screens 를 활용하는 비즈니스 로직이 있고,  screens 자체는 프러덕션 코드에서 참조하고 있지 않으므로 private 으로 캡슐화해줘도 좋겠네요!

Hogu59 replied:
*) Screen 객체 명은 명시적 표현을 위해 ScreeningMovie로 변경하였습니다!

일단 `DummyTheaterList`에서  영화 아이디로 극장 리스트를 찾는 함수에서 
```kotlin
override fun findTheaterList(movieId: Long): List<Theater>   {
  return list.filter { theater ->
      theater.screeningMovies.any { screen ->
          screen.movieId == movieId
  }
}
```
와 같은 방식으로 `(구)screens (현)screeningMovies` 에 직접 접근하고 있었는데, 캡슐화에 대해 말씀해주시니 확실히 가진 정보들을 가급적이면 은닉할 수 있도록 하는 방식이 조금 더 좋겠다는 생각이 드네요!

그런 맥락에서 `screens`를 private 화 했고, 문제가 되었던 해당 함수도 `screens`에 직접 참조하지 않도록 코드를 변경해두었습니다!
```kotlin
override fun findTheaterList(movieId: Long): List<Theater>   {
    return list.filter { theater ->
        theater.findScreenScheduleListWithMovieId(movieId).isNotEmpty()
    }
}
```

가이드 주셔서 감사합니다!

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/homefragments/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    override fun ticketingButtonClick(movieId: Long) {
+        val bottomSheet = TheaterBottomDialogFragment(DummyTheaterList, movieId)
```
모델에 대한 source 는 presenter 에서 불러와야 하지 않을까요?

현재와 같은 방식은 MovieListFragment  와 TheaterBottomDialogFragment 가 동일한 생명주기 내에서 공유하고 싶은 데이터가 있을때만 의미있는 구현이 될 것 같아요.

Hogu59 replied:
혹시 문제가 되는 부분이
1. DummyTheaterList 를 Fragment(View) 가 몰라야 한다 의 의미일까요..?
2. 아니면 사실상 presenter에서 구현되어야 할 onTicketingButtonClicked()의 내용이 view에서 전부 수행되어 문제가 되었다는 의미일까요?  ( = bottomSheet 의 생성은 프레젠터에서 진행되어야한다.)

1번의 경우라면, TheaterListRepository로 추상화 한 이유 자체가 추후 실제로 사용될 TheaterList가 DummyList가 아닐 것을 대비해서 작성한 것이므로 해당 위치에서 사용하는 것이 presenter의 외부 의존성을 제거할 수 있는 방식이 되지 않을까 싶었습니다.

2번의 경우라면 충분히 맞는 말씀이라고 생각합니다. bottomSheetDialogFragment의 생성, 즉 객체의 생성 자체는 프레젠터에서 수행하고 적절히 프레젠터와 뷰의 역할을 분리하는 것이 맞는 것 같습니다.

해당 내용은 아래와 같이 수정해보았습니다!
`MovieListFragment class`
```Kotlin
override fun ticketingButtonClickListener(movieId: Long) {
    presenter.onTicketingButtonClick(DummyTheaterList, movieId)
}

override fun showBottomSheetFragment(bottomSheetDialogFragment: BottomSheetDialogFragment) {
    bottomSheetDialogFragment.show(childFragmentManager, bottomSheetDialogFragment.tag)
}
```

`MovieListPresenter`
```kotlin
override fun onTicketingButtonClick(
    theaterListRepository: TheaterListRepository,
    movieId: Long,
) {
    val bottomSheetFragment = TheaterBottomDialogFragment(theaterListRepository, movieId)
    view.showBottomSheetFragment(bottomSheetFragment)
}
```

로 수정해보았습니다! 제가 의도하신 바를 제대로 이해했는지 알 수 있을까요? 🤔

vagabond95 replied:
코멘트 내용이 조금 모호했던거 같네요 ^^;

제가 의도했던 바는 "`TheaterBottomDialogFragment` 를 생성할 때, 꼭 theaterListRepository 를 주입받아야 하는가?" 였습니다.
TheaterBottomDialogPresenter 내부에서 초기화를 시키면 안되는 것일까요~?
다른 Repository 는 Presenter 에서 초기화를 시켜주셨는데, 왜 해당 Repository 만 fragment 를 통해 주입받도록 처리하셨는지가 궁금했습니다.

Hogu59 replied:
이부분에 대해서는 theaterListRepository 를 주입하는 쪽이 다르다보니 조금 이상하게 느껴질수도 있겠네요.
항상 presenter 생성시에 fragment에서 `DummyTheaterList`를 주입하는데, 이번에는 Fragment 생성시에 주입하고 있으니 일관적이지 않게 느껴질수 있겠네요.

MovieList -> BottomFragment의 흐름 자체가 하나의 흐름이라고 생각해서 그렇게 설계했던 것 같습니다.
그렇게 되면 추후에 DummyTheaterList가 아닌 실제 TheaterList를 사용한다고 할 때, MovieListFragment에서 주입하는 theaterListRespository 만 수정함으로써 더 많은 범위의 수정을 커버할 수 있다고 생각했습니다.

흠... 근데 비슷한 맥락으로 생각해보면, `그렇다면 왜 앞의 액티비티나 fragment에서는 TheaterListRepository를 주입하지 않나요? 그런식으로 하면 맨 처음의 액티비티 혹은 프라그먼트에서 한번 수정함으로써 모든 수정을 커버할 수 있을텐데요?` 라는 자문을 할 때, 답을 하기가 어려워지네요...

일단 제리가 가이드 해준 방향대로 TheaterBottomDialogFragment는 theaterListRepository를 주입받지 않는 형태로 변경했습니다!

다만 궁금한 점은 이런 식으로 코드를 작성한다면 유지보수성은 떨어지는 것이 아닌가라는 생각이 듭니다.
예를 들어, `안드로이드 개발자님, 이제 실제 DB가 연결되었습니다! 그동안 DummyList를 사용하셨던 것 같은데, 실제 DB로 연결해서 사용하시면 됩니다. 아래 첨부된 쿼리로 검색하시면 DummyList에서 사용하시던 형식과 동일하게 자료를 뽑아 사용하실 수 있습니다.` 라는 연락을 받는다면, 현재의 코드에서는 DummyTheaterList를 사용한 모든 코드를 검색해서 수작업으로 수정해야하는데, 유지보수성을 생각할 때 좀 더 유용한 방법이 혹시 있는지 알 수 있을까요? 🤔 

p.s. 예시를 적고보니 말투가 꽤나 헤드퍼스트 시리즈의 예시문 같네요....ㅋㅋㅋㅋ

vagabond95 replied:
얘기해주신 부분을 하나씩 살펴볼게요 :)

>  MovieListFragment에서 주입하는 theaterListRespository 만 수정함으로써 더 많은 범위의 수정을 커버할 수 있다고 생각했습니다.

BottomFragment 가 한군데서만 사용된다면 효용이 있을 수 있으나, 그 이상 사용이 된다면 의미가 퇴색될 것 같아요. 그리고, BottomFragment 에서 어떠한 repo 를 통해 데이터를 불러올지는 BottomFragment 스스로 결정하는게 더 변경에 유연하지 않을까요~? 사용하는 쪽에서 신경을 쓰게되면, 사용하는 곳이 늘때마다 변경포인트도 늘어나게 되니까요 ㅎㅎ

> 현재의 코드에서는 DummyTheaterList를 사용한 모든 코드를 검색해서 수작업으로 수정해야하는데, 유지보수성을 생각할 때 좀 더 유용한 방법이 혹시 있는지 알 수 있을까요?

유지 보수성을 고려하는 측면, 너무 좋다고 생각합니다! 현 구조에서는 DummyTheaterList -> 서버 데이터로 변경을 시도할 경우 모든 구현체를 바꿔줘야 하는 것이 맞습니다. 

다만 이 부분은 다음과 같은 방법으로 개선이 될 수 있어보여요.
1. Dependency Injection 라이브러리 활용
2. TheaterListRepository 구조 개선

1번의 경우, 현 단계에서는 내용을 벗어나는 내용이므로 키워드 정도만 알아두셔도 좋습니다. 의존성을 주입하는 것을 자동으로 처리하도록 도와주는 도구라고 보시면 될 것 같아요.

2번의 경우, 다음과 같이 리펙토링 해보면 어떨까요?
```
object TheaterListRepositoryImpl : TheaterListRepository {
    // todo 추후에 서버 데이터로 교체
    val dummyLocalList = DUMMY_LIST

    override val list: List<Theater> =  dummyLocalList

    override fun findTheaterList(movieId: Long): List<Theater> {
        return dummyLocalList.filter { theater ->
            theater.findScreenScheduleListWithMovieId(movieId).isNotEmpty()
        }
    }

    override fun findTheaterOrNull(theaterId: Long): Theater? {
        return dummyLocalList.find { theater ->
            theater.id == theaterId
        }
    }
}
```

Dummy 데이터만을 위한 repo 구현체를 만드는 것이 아니라, repo 구현체 자체는 실제 동작과 유사하게 만들고 data source 만 더미 데이터로 가지고 있는 구조인데요.

이러면, TheaterListRepositoryImpl 를 참조하는 쪽에서는 변경사항이 발생하지 않고 추후에 서버 작업만 해당 repo 내에서 진행해주주면 되겠죠 :)

인터페이스 분리 및 재사용성을 잘 고려해주셨는데, 조금만 더 케이스를 확장시켜보면 다음과 같이 풀수도 있다는 예시를 보여드리고자 남겼습니다~!

vagabond95 reviewed (changes requested):
악어 안녕하세요! 미션 진행동안 잘 부탁드립니다 :)

전반적으로는 크게 코멘트 드릴 부분이 보이지 않았던 것 같습니다 👍 
커밋 범위를 지정해주신 것도 좋았구요 ㅎㅎ

극장 선택, Fragment 추가 등 해당 단계의 요구사항을 중심으로 리뷰진행하였습니다.
확인 후 리뷰요청 부탁드릴게요~!

Hogu59 committed:
rename(ScreeningMovie): change class naming signature
- Prior naming signature is unclear to find out the meaning
- before: Screen
- after: ScreeningMovie

Hogu59 committed:
rename(tablelayout): avoid using abbreviation
- before: tl_seats
- after: tablelayout_seats

Hogu59 committed:
rename(BindingImageAdapter): avoid using abbreviation
- before: imgRes, resId
- after: imgResource, resourceId

Hogu59 committed:
refactor(MovieAdapter): erase un-using variable

Hogu59 committed:
rename(seat_selection.xml): change data binding naming
- apply camel case
- before: seating_system
- after: seatingSystem

Hogu59 committed:
refactor(HomeActivity): Optimize activity
- Set-up fragments and re-use
- use `commit { }` for set-up initial fragment

Hogu59 committed:
refactor(Seat): use extension function
- To get a String type of Seat,
- before: Use Utils package
- after:
 - Use extension function
 - Function definition was moved to Seat domain logic
 - Delete Utils package

Hogu59 committed:
refactor(Theater): change access modifier of property
- screnningMovies: public -> private

Hogu59 committed:
refactor(HomeActivity): optimize replacing fragments
- Delete `addToBackStack` as it is un-necessary
- Add a tag(String) to replace method of fragmentTransaction
*) Disable Dark-mode

Hogu59 committed:
refactor(MovieListFragment): apply mvp pattern
- delete `ticketingButtonClick` method and separate it by applying mvp pattern,
- to show bottomSheetDialogFragment,use childFragmentManager instead of parentFragmentManager
- to check lifecycles, add Log lines

Hogu59 committed:
test(presenter): write a few test codes

Hogu59 commented:
제리 안녕하세요!
악어입니다!

가이드 주신 내용과 추가적으로 코드 수정을 수행해봤습니다!
확인해주시면 감사하겠습니다!

특히 리뷰주신 부분 중 childFragmentManager / parentFragmentManager 중 하나를 인자로 넘겨주는 부분에 대해서 공부해보고 로그를 찍어서 확인해보았지만 아직 감 잡기가 조금 어려운 것 같습니다.. 혹시 조금만 더 가이드를 주실 수 있을까요..?
https://github.com/woowacourse/android-movie-theater/pull/65#discussion_r1588817888

작업 내역 
https://github.com/Hogu59/android-movie-theater/compare/3967d9e..hogu59

감사합니다!

vagabond95 commented on app/src/main/res/layout-land/activity_seat_selection.xml:
```diff
@@ -0,0 +1,267 @@
+
+                    <androidx.appcompat.widget.AppCompatTextView
+                        android:id="@+id/tv_seat_item_0"
+                        style="@style/seatItemStyle"
```
style 활용 👍

vagabond95 commented on app/src/main/res/layout/item_theater.xml:
```diff
@@ -0,0 +1,94 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toStartOf="@id/item_theater_name_tv"
+            app:layout_constraintTop_toBottomOf="@id/item_theater_name_tv"
+            android:text="@{String.valueOf(theater.findScreenScheduleListWithMovieId(movieId).size())}"
```
xml (= View) 에서 Model 에 대한 세부 사항을 너무 많이 알고 있는 것 같아요.
view 를 그리기 위한 데이터만 전달하도록 리펙토링 해볼 수 있을까요?

item 을 그리기 위한 별도 UiModel 을 두어도 괜찮겠습니다 :)

Hogu59 replied:
Ui model로 ScreeningTheater를 따로 만들어보았습니다.
presenter에 `movieId(Long)` 를 넘겨 주었을 때, List<ScreeningTheater>를 생성하여 view 로 넘겨주어 필요한 정보만 출력할 수 있도록 해두었습니다!

UiModel을 처음 사용해보는데, 이런 방식으로 사용해도 괜찮은지, 혹은 더 좋은 방법이 있는지 알려주시면 감사하겠습니다!

가이드 주셔서 감사합니다 😄

vagabond95 replied:
잘 구현해주셨습니다! UiModel 은 어떠한 best practice 가 있는 것은 아니고, `계층간 분리 및 UI 표현`이라는 목적을 달성한다면 어떤 형태로든 좋다고 생각합니다. ㅎㅎ

vagabond95 replied:
하지만 여기서 끝내면 아쉬우니, 한가지 저의 스타일을 말씀드려보면

가령, 특정 UI 에서 formatting 을 해야하는 스펙이 있다면 2가지 접근 방법이 있습니다.
1. view 내에서 상태를 가지고 와서 formatting 처리를 하기
2. UiModel 내에서 formatting 한 값을 전달해주기

후자로 진행된다면, View 에서는 formatting 에 대한 정보가 캡슐화되므로 좀 더 변경에 유연하게 대응할 수 있다는 장점이 있습니다 :)

vagabond95 commented on app/src/main/java/woowacourse/movie/model/Movie.kt:
```diff
@@ -0,0 +1,14 @@
+    val runningTime: Int,
+    val introduction: String,
+) {
+    fun screenPeriodToString(): String {
```
View 에서 시간을 표기하기 위한 함수로 보이는데요,
해당 모델은 도메인 계층에 위치하고 있으므로 해당 함수가 선언되어있는 것이 조금 어색해보입니다.

- extension function
- 별도 UiModel 로 분리

등의 방법으로 개선이 필요해보여요~!

Hogu59 replied:
위에서는 UiModel 로 분리해보고, 이번에는 사용되는 많은 곳에서 Movie 객체의 대부분의 정보를 필요로 하다보니 확장함수를 사용해봤습니다.
둘 다 사용해보니 추후 사용성을 생각할 때, 확장함수보다는 UiModel을 따로 두는 것이 조금 더 바람직할 것 같다는 생각이 드네요!

두 가지 방법 같이 추천해주셔서 감사합니다! 😄

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/homefragments/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,72 @@
+        savedInstanceState: Bundle?,
+    ) {
+        super.onViewCreated(view, savedInstanceState)
+        // Log.d("MovieListFragment", "onViewCreated")
```
테스트가 끝난 로그들은 모두 지워주시면 좋겠습니다~!

Hogu59 replied:
삭제처리 하겠습니다!

vagabond95 commented on app/src/main/java/woowacourse/movie/presentation/homefragments/movieList/MovieListPresenter.kt:
```diff
@@ -0,0 +1,23 @@
+        movieId: Long,
+    ) {
+        val bottomSheetFragment = TheaterBottomDialogFragment(theaterListRepository, movieId)
+        view.showBottomSheetFragment(bottomSheetFragment)
```
제가 코멘트를 모호하게 드려서인 탓이 크겠지만 😅 
Presenter 에서 Fragment 를 생성하는 일은 지양해야 할 것 같아요!

Hogu59 replied:
https://github.com/woowacourse/android-movie-theater/pull/65#discussion_r1591911368

위에 답변 적어두었습니다!

알려주셔서 감사합니다 😄

vagabond95 reviewed (changes requested):
모호한 코멘트 하나가 있어서 혼동이 오셨을 것 같네요 🥲 질문 주신 내용에 대한 참고할 수 있는 자료를 첨부드렸으니 참고 부탁드려요~!

추가 코멘트 확인 후 리뷰요청 부탁드리겠습니다 :)

Hogu59 committed:
remove: delete un-using test codes
- Log codes to check fragments' life-cycle

Hogu59 committed:
feat(ScreeingTheater): create an ui-model
- A ui-model to only send necessary information
- contains 4 properties, movieId(Long), theaterId(Long), theaterName(String), numberOfScreeningSchedule(Int)

Hogu59 committed:
feat(Movie): remove a method made for view
- `screenPeriodToString`: A method to get a string type of startDate to endDate of screeningDates
- As it is a domain and the method is only used for view, delete method from domain and use extension functions for same usage.

Hogu59 committed:
refactor(TheaterBottomDialogFragment): remove fragment dependency
- Before: TheaterBottomDialogFragment contains TheaterListRepository as constructor parameter.
- After: TheaterBottomDialogFragment sends DummyTheaterList to presenter without injection from MovieListFragment

Hogu59 commented:
제리 안녕하세요!

이번 리뷰 덕분에 UiModel을 처음 만들어서 사용해보았는데,
1) 필요한 정보만 사용할 수 있다는 점
2) xml에서 필요한 정보에만 접근할 수 있다는 점
에서 유용성을 많이 느꼈던 것 같습니다!

이외에도 `TheaterListRepository` 를 주입한 내용에 대해서 가이드 주신 대로 수정했는데 동시에 드는 의문이 생겼는데, 해당 내용에 대해서 의견을 조금 들을 수 있을까요..?

감사합니다!

vagabond95 reviewed (approved):
재사용성에 대해서 평소에도 깊은 고민을 하시는 것 같네요 👍 

repo 와 ui model 에 대한 간략한 제 생각을 남겼는데, 도움이 되셨길 바랍니다 :)
다음 미션 진행을 위해 1,2 단계는 여기서 머지할게요!

마무리까지 화이팅입니다!
